api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service that are externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of ExternalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// One or more Service types can be specified, if desired.
+	//
+	// If no service types are provided, ExternalDNS will be
+	// configured to create DNS records for LoadBalancer services
+	// only by default.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={"LoadBalancer"}
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	ServiceType []corev1.ServiceType `json:"serviceType,omitempty"`
+}
+
+type ServicePublishPolicy string
+
+// +kubebuilder:validation:Enum=PublishExternal;PublishInternalExternal
+const (
+	ServicePublishPolicyExternalOnly        ServicePublishPolicy = "PublishExternal"
+	ServicePublishPolicyExternalAndInternal ServicePublishPolicy = "PublishInternalExternal"
+)
+
+type ExternalDNSCRDSourceOptions struct {
+	// Kind is the kind of the CRD
+	// source resource type to be
+	// consumed by ExternalDNS.
+	//
+	// e.g. "DNSEndpoint"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Kind string `json:"kind"`
+
+	// Version is the API version
+	// of the given resource kind for
+	// ExternalDNS to use.
+	//
+	// e.g. "externaldns.k8s.io/v1alpha1"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Version string `json:"version"`
+
+	// LabelFilter specifies a label filter
+	// to be used to filter CRD resource instances.
+	// Only one label filter can be specified on
+	// an ExternalDNS instance.
+	//
+	// +optional
+	LabelFilter map[string]string `json:"labelFilter,omitempty"`
+}
+
+// ExternalDNSStatus defines the observed state of ExternalDNS.
 type ExternalDNSStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Conditions is a list of operator-specific conditions
+	// and their status.
+	Conditions []ExternalDNSOperatorCondition `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the most recent generation observed.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Provider is the configured provider in use by ExternalDNS.
+	Provider ExternalDNSProvider `json:"provider,omitempty"`
```

Comment:
Why is this in status?  I'm betting you reject changes to it, but explain.

Reply:
We were planning on using this status field to reflect the currently configured provider, so a user would know if their provider credentials are valid.

Reply:
If credentials are invalid, that can be reflected in status conditions.  

I had in mind that `spec.provider` could be omitted, in which case the operator would supply defaults and use `status.provider` field to communicate those defaults.  However, my current thinking is that the operator may create a default ExternalDNS CR with `spec.provider` filled in, so defaulting just the field may be unnecessary, and I see `spec.provider` is required, so maybe `status.provider` can be omitted.

Reply:
My mistake. 
For now I will omit `status.provider`. Thanks!